identity-android/src/main/java/com/android/identity/android/direct_access/DirectAccess.kt

```diff
@@ -0,0 +1,493 @@
+        return asn1UtcTime.encoded
+    }
+
+    /**
+     * Creates a credential at the given slot.
+     *
+     * This is called during initialization of a [DirectAccessCredential], so there is no need to call
+     * this explicitly when dealing with DirectAccessCredentials.
+     *
+     * Note: this operation will not alter the state of the applet, ie. would not swap out the
+     * existing credential in the slot.
+     *
+     * @param documentSlot the slot to provision a credential in
+     * @return signingKeyCertificate, encryptedPresentationData
+     */
+    fun createCredential(
+        documentSlot: Int
+    ): Pair<X509CertChain, ByteArray> {
+        val response: ByteArray?
+        try {
+            transport.closeConnection()
+            transport.openConnection()
+
+            val scratchpad = ByteArray(2)
+            val bos = ByteArrayOutputStream()
+            // set instruction
+            setShort(scratchpad, 0, CMD_MDOC_CREATE_PRESENTATION_PKG.toShort())
+            bos.write(scratchpad)
+            bos.write(documentSlot)
+
+            // todo @venkat the validity duration should be set during certification as opposed to creation
+            val now = Clock.System.now()
+            val notBeforeBytes = encodeValidityTime(now)
+            val notAfterBytes = encodeValidityTime(now + Duration.parse("3600d"))
+
+            // Set Not Before
+            setShort(scratchpad, 0, notBeforeBytes.size.toShort())
+            bos.write(scratchpad)
+            bos.write(notBeforeBytes)
+
+            // Set Not After
+            setShort(scratchpad, 0, notAfterBytes.size.toShort())
+            bos.write(scratchpad)
+            bos.write(notAfterBytes)
+            val apdu = makeCommandApdu(bos.toByteArray())
+
+            response = transport.sendData(apdu)
+            val status = getAPDUResponseStatus(response)
+            check(APDU_RESPONSE_STATUS_OK == status) {
+                "createPresentationPackage failed. Response status: $status" }
+            val input = response.copyOf(response.size - 2)
+
+            var signingCert: X509CertChain? = null
+            var encryptedData: ByteArray? = null
+            val map = decode(input)
+            val keys = map.asMap.keys
+            for (keyItem in keys) {
+                val value = keyItem.asNumber.toInt()
+                when (value) {
+                    KEY_CERT -> {
+                        signingCert = X509CertChain(listOf(X509Cert(map[keyItem].asBstr)))
+                    }
+
+                    KEY_ENC_DATA -> {
+                        val encBytesItem = map[keyItem]
+                        encryptedData = encBytesItem.asBstr
+                    }
+
+                    else -> throw IllegalStateException("createPresentationPackage unknown key item")
+                }
+            }
+            transport.closeConnection()
+            return Pair(signingCert!!, encryptedData!!)
+        } catch (e: IOException) {
```

Comment:
In case of any other type of exception, transport.closeConnection() would not be called. Should that be moved to a finally block?